cli/server.go

```diff
@@ -944,15 +944,27 @@ func (r *RootCmd) Server(newAPI func(context.Context, *coderd.Options) (*coderd.
 			var provisionerdWaitGroup sync.WaitGroup
 			defer provisionerdWaitGroup.Wait()
 			provisionerdMetrics := provisionerd.NewMetrics(options.PrometheusRegistry)
-			for i := int64(0); i < vals.Provisioner.Daemons.Value(); i++ {
+
+			// Create a list of daemon types. The length is the total number of built-in provisioners, and
```

Comment:
So, at a high level, _provisioner daemons_ don't have a type.  _Provisioners_ have a type, and a provisioner daemon can accept multiple provisioner types via the `Connector` map.

Do we really want/need to have a different number of _daemons_ for echo and terraform?  Or the choice really between: echo, terraform, or both?  If so, maybe it makes sense to keep the flag about the number as is, and then add a flag (default true) that controls whether to include terraform provisioners.

Reply:
> So, at a high level, provisioner daemons don't have a type. Provisioners have a type, and a provisioner daemon can accept multiple provisioner types via the Connector map.

Yes, but that assumes the provisioner implementation can support both types. Right now we have an echo provisioner daemon implementation, and a terraform. To unify those into a single daemon, we'd need a daemon implementation that can support both.

I think in practice, having provisioner daemons support 1 job type is reasonable if those jobs are unique enough. Example being if we add OpenTF as a new type, you could probably make the `terraform` provisioner support both `terraform` and `open-tf`.

----

For the flag. The status quo is you can have either terraform provisioners **xor** echo provisioners. For e2e testing, it makes sense to have both, as the echo provisioner obviously has less overhead for some of our tests.

I think it makes sense to be able to control whether you run terraform only (prod), both (e2e), just echo (unit testing). So the flags need to support these 3 cases.

----

The `--provisioner-daemons` flag is the flag our customers use today to control the number of built-in terraform provisioners. Given `--provisioner-daemons-echo` is hidden and only used in testing, it feels reasonable to give it the same functionality as it's production counterpart, just for echo provisioners instead.